cluster-autoscaler/loop/trigger.go

```diff
@@ -43,38 +45,59 @@ type scalingTimesGetter interface {
 	LastScaleDownDeleteTime() time.Time
 }
 
+// provisioningRequestProcessingTimesGetter exposes recent provisioning request processing activity regardless of wether the
+// ProvisioningRequest was marked as accepted or failed. This is because a ProvisioningRequest being processed indicates that
+// there are other ProvisioningRequests that require processing regardless of the outcome of the current one. Thus, the next iteration
+// should be started immediately.
+type provisioningRequestProcessingTimesGetter interface {
+	LastProvisioningRequestProcessTime() time.Time
+}
+
 // LoopTrigger object implements criteria used to start new autoscaling iteration
 type LoopTrigger struct {
-	podObserver        *UnschedulablePodObserver
-	scanInterval       time.Duration
-	scalingTimesGetter scalingTimesGetter
+	initialized                          bool
+	podObserver                          *UnschedulablePodObserver
+	scanInterval                         time.Duration
+	scalingTimesGetter                   scalingTimesGetter
+	provisioningRequestProcessTimeGetter provisioningRequestProcessingTimesGetter
 }
 
 // NewLoopTrigger creates a LoopTrigger object
-func NewLoopTrigger(podObserver *UnschedulablePodObserver, scalingTimesGetter scalingTimesGetter, scanInterval time.Duration) *LoopTrigger {
+func NewLoopTrigger(scalingTimesGetter scalingTimesGetter, provisioningRequestProcessTimeGetter provisioningRequestProcessingTimesGetter, scanInterval time.Duration) *LoopTrigger {
 	return &LoopTrigger{
-		podObserver:        podObserver,
-		scanInterval:       scanInterval,
-		scalingTimesGetter: scalingTimesGetter,
+		podObserver:                          nil,
+		scanInterval:                         scanInterval,
+		scalingTimesGetter:                   scalingTimesGetter,
+		provisioningRequestProcessTimeGetter: provisioningRequestProcessTimeGetter,
 	}
 }
 
+// Initialize initializes the LoopTrigger object by providing a pointer to the UnschedulablePodObserver
+func (t *LoopTrigger) Initialize(podObserver *UnschedulablePodObserver) {
```

Comment:
The aim was to make minimal changes to the args and return values of the `buildAutoscaler` function.

The trigger can only be initialized within the `buildAutoscaler` function as the injector is present there. @yaroslava-serdiuk felt that creating the injector in the `run` function and passing that to `buildAutoscaler` was not good because the injector is only a relevant component for CA if the user has ProvisioningRequests enabled and thus should not be a part of the `buildAutoscaler` args.

The `podObserver` can only be created within the `run` function as it requires the background context of the function. Therefore, having an initialize method which serves as a setter for the `podObserver` was deemed as the best solution.

Reply:
Thanks for sharing the background!

I think the context can safely be created before the call to `buildAutoscaler` and passed there - you can then remove two-phase init and actually simplify `podObserver` creation a bit too by reusing autoscaling options available there.

Reply:
Modified accordingly.